common/web/lm-worker/src/transformUtils.ts

```diff
@@ -0,0 +1,25 @@
+class TransformUtils {
+  static isWhitespace(transform: Transform): boolean {
+    // Matches prefixed text + any instance of a character with Unicode general property Z* or the following: CR, LF, and Tab.
+    let whitespaceRemover = /.*[\u0009\u000A\u000D\u0020\u00a0\u1680\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u200b\u2028\u2029\u202f\u205f\u3000]/i;
+
+    // Filter out null-inserts; their high probability can cause issues.
+    if(transform.insert == '') { // Can actually register as 'whitespace'.
+      return false;
+    }
+
+    let insert = transform.insert;
+
+    insert = insert.replace(whitespaceRemover, '');
+
+    return insert == '';
+  }
```

Comment:
This seems a little convoluted.

* I'm really not sure exactly what we are testing here. It seems like you are looking for `insert.match(/[\u0009...]$/)`? Is it intended to match having a single whitespace character on the end of the text? If so, then the function name is not very clear. If not, then AFAICT the function is not correct.

* `i` modifier should not be necessary on the regex -- there's no case to be insensitive about 🤣.

* Can we use a [character class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Unicode_Property_Escapes) for the regex match rather than a set of characters? (e.g. `insert.match(/[\p{Z}\r\n]$/u)`. Note Chrome 50+ though (so we could have the shorter regex as a comment if we can't use it for back-compat reasons).

* Finally, can we have a javadoc comment rather than the in-function comment?

Reply:
Keep in mind that I did not edit this function _at all_.  I simply extracted it from its prior home.  From `master`:

https://github.com/keymanapp/keyman/blob/80d7ce78bedccda20f73982ac74d26da73b5c63c/common/web/lm-worker/src/model-compositor.ts#L19-L33

This all dates back to #1851, and @eddieantonio was the one responsible for that regex: https://github.com/keymanapp/keyman/pull/1851#discussion_r300022560

Can't speak to why he didn't use a character class here, but I figure that there was some reason or other.  For what it's worth, following that character class link and drilling down to the actual classes:  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Character_Classes

I don't see a compatibility chart there.  Now, there are other ways to search for this...

https://caniuse.com/mdn-javascript_builtins_regexp_property_escapes

Note the minimum Chrome version listed there - if desktop Chrome didn't support these property escapes before version 64, I doubt that mobile had it implemented at that time.  This, to me, implies a strong chance that using them would break predictive text outright on non-updated Android 5.0 devices, at the least.

------

Where do you see a `$` in the regex string here?  Is it implied with the flags on the pattern?  It's certainly not present in plain-text.

Reply:
So... feel free to make an issue about the function feeling weird to you, but since there are no edits to this function within this PR, any potential changes should be considered out of scope here.

Reply:
Yep, happy to postpone fixing it. Wonder if this function may be causing issues though because it seems weird...

> Where do you see a `$` in the regex string here? Is it implied with the flags on the pattern? It's certainly not present in plain-text.

It's implied by `return insert == '';` -- so this will only be true if the regex deletes everything in the string, which means it must be the last char in the string.

Reply:
Keep in mind that we're dealing with incoming transforms - in essence, this is used on incoming keystrokes, not existing context.  (Hence the classname of `TransformUtils`.  (Emphasis on **Transform**)

```typescript
isWhitespace({
  insert: 'a ',
  deleteLeft: 0
)
```

will yield `false` - after all, that 'a' will remain.

Meanwhile, this one:

```typescript
isWhitespace({
  insert: '\u{0020}\u{0020}\u{0020}\u{0020}', // four spaces in a row; doesn't format well
  deleteLeft: 0                               // on web without the \u stuff
)
```

will yield `true`.  The design is meant to check if any prediction-triggering keystroke is _fully_ whitespace or not.

Its uses do imply a few assumptions, though - we're not doing any `hasWhitespace` checks, which might be more valid for niche, likely-unintuitive scenarios.

Reply:
Are you sure? I just ran this:

```ts
let foo = [
  'a ',
  ' ',
  '   ',
  ' x x ',
  'xxx   ',
  'xxx'
];

for(let f of foo) {
  console.log(`"${f}": `, isWhitespace(f));
}
```

and the result was:

```
$ node foo.js
"a ":  true
" ":  true
"   ":  true
" x x ":  true
"xxx   ":  true
"xxx":  false
```

Reply:
If it is supposed to check if a given transform is 100% whitespace and not empty, then:

```js
let whitespace = /^[\u0009\u000A\u000D\u0020\u00a0\u1680\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u200b\u2028\u2029\u202f\u205f\u3000]+$/; 

return transform.insert.match(whitespace) !== null;
```

Reply:
Oh right... regex matching by default gets _the portion_, rather than checking against the whole.  I tend to forget that sometimes.

Reply:
Sorry on the delay.  Just got back to this.

Though, I did go ahead and add unit testing for the function while I was at it.